proposals/v3/api.md

```diff
@@ -0,0 +1,1119 @@
+      "type:led",
+      "led"
+    ]
+  }
+]
+```
+
+**Fields**
+
+| Field | Description |
+| ----- | ----------- |
+| *id* | The globally unique ID for the device. |
+| *info* | A human readable string providing identifying info about a device. |
+| *type* | The type of the device, as defined by its plugin. |
+| *plugin* | The ID of the plugin which the device is managed by. |
+| *tags* | A list of the tags associated with this device. One of the tags will be the `id` tag. |
+
+##### Error (500, 400)
+See: [Errors](#errors)
+
+
+---
+
+### Tags
+```
+GET http://HOST:5000/synse/v3/tags
+```
+
+List all of the tags currently associated with devices.
+
+This will list the tags in the specified tag namespace. If no tag namespace
+is specified (via query parameters), the default tag namespace is used.
+
+By default, this endpoint will omit the `id` tags since they match the
+device id enumerated by the [`/scan`](#scan) endpoint. The `id` tags can
+be included in the response by setting the `ids` query parameter to `true`.
+
+Multiple tag namespaces can be queried at once by using a comma delimiter
+between namespaces in the `ns` query parameter value string, e.g.
+`?ns=default,ns1,ns2`.
+
+Tags are sorted alphanumerically.
+
+#### Query Parameters
+
+| Key | Description |
+| --- | ----------- |
+| ns | The tag namespace(s) to use when searching for tags. (default: `default`) |
+| ids | A flag which determines whether `id` tags are included in the response. (default: `false`) |
+
+#### Response
+
+##### OK (200)
+```json
+[
+  "default/tag1",
+  "default/type:temperature"
+]
+```
+
+##### Error (500, 400)
+See: [Errors](#errors)
+
+
+---
+
+### Info
+```
+GET http://HOST:5000/synse/v3/info/{device_id}
+```
+
+Get the full set of metainfo and capabilities for a specified device.
+
+
+#### URI Parameters
+
+| Parameter | Required | Description |
+| --------- | -------- | ----------- |
+| *device_id* | yes | The ID of the device to get info for. IDs should be globally unique. |
+
+#### Response
+
+##### OK (200)
+```json
+{
+  "timestamp": "2018-06-18T13:30:15.6554449Z",
+  "id": "34c226b1afadaae5f172a4e1763fd1a6",
+  "type": "humidity",
+  "kind": "humidity",
+  "metadata": {
+    "model": "emul8-humidity"
+  },
+  "plugin": "emulator plugin",
+  "info": "Synse Humidity Sensor",
+  "tags": [
+      "type:humidity",
+      "humidity",
+      "vio/fan-sensor"
+  ],
+  "capabilities": {
+    "mode": "rw",
+    "read": {},
+    "write": {
+      "actions": [
+        "color", 
+        "state"
+      ]
+    }
+  },
+  "output": [
+    {
+      "name": "humidity",
+      "type": "humidity",
+      "precision": 3,
+      "scaling_factor": 1.0,
+      "unit": {
+        "name": "percent humidity",
+        "symbol": "%"
+      }
+    },
+    {
+      "name": "temperature",
+      "type": "temperature",
+      "precision": 3,
+      "scaling_factor": 1.0,
+      "unit": {
+        "name": "celsius",
+        "symbol": "C"
+      }
+    }
+  ]
+}
+```
```

Comment:
I think it may be useful to also include a way to specify some kind of device kind description. Not sure if others see this as valuable, so feedback is welcome. I can think of two ways of doing this:

* `ref`: a URL to a page/somewhere which provides additional information on the device type. this could link to a GitHub readme/wiki, etc. It would allow a more verbose description for a user to read up on its usage, what it provides, what actions it may support, etc.

* `description`: similar to the above, but instead of a link, it is just a block of text directly in the response. it can be shorter than the description that is linked to, but there is no real enforcement of what this could contain. having it directly in the response means that the API consumer doesn't need to resolve another link in order to get to the information.

I'd probably favor `description` because it makes the information more easily accessible, it doesn't rely on the other link to work, the contents are better-known (e.g. a link could point to anything..).

Reply:
A client case for synse (how auto_fan uses it since OpenDCRE v1):
/synse/version
/synse/scan (possibly force)
Info is part of the configuration.
auto_fan identifies sensors via scan and info from the scan results.

Info is the configured field for client purposes. It's how the client ids a specific sensor.

Why do we need a new description field? What is the client side use case? Why do we need to support it?

Reply:
The client side use case that I was envisioning was being able to specify info about the device that may be useful for a customer to know, e.g. when it was installed or replaced last, a description of where it exists (which we kinda use it for now), if it supports writing, what actions it supports along with descriptions of what those actions do. it can also say what kind of data it expects to have in the `data` field, since that isn't really easy to encode anywhere in a meaningful way, e.g.

```
Wedge 1 LED

This is the LED strip at the top of the Chamber wedge. It was last serviced
on 1/1/2019.

The following actions and corresponding data are supported on write:
ACTION   DESC                           DATA
state    The power state of the LED.    'on', 'off', 'blink'
color    The color of the LED bar.       A hexadecimal color value between 000000 and FFFFFF
```

It may not be tenable to maintain this info for every device, any maybe we wouldn't want to. This was just along the lines of my original thought, particularly that a user may not know what restrictions there are on a device for writing. They could always go to the documentation page for the plugin, but that didn't seem ideal.

It could be a moot point, since a lot of the interaction is not done directly by the user but by something that knows the restrictions (e.g. auto-fan, UI), and as long as the error responses for invalid write data is descriptive enough a user can figure it out.

Reply:
Currently, the `info` field is the only way the frontend knows how to associate a Lock, with a Locker. It's not fully in place yet, we need to add this information to the plugin configuration. Until tag information is available to associate a device with a locker, the `info` field needs to stay succinct, and consumable. It's actually an available filter in their GraphQL queries for device. We allow filtering based on substrings in device info fields. I am **for** the idea of a field that returns information regarding how to interact with a device. But it looks like that's already contained in the `capabilities` field. The human readable device capabilities is a nice to have, but it certainly should not go in the `info` field.

Reply:
Yeah, none of the proposed changes here should change anything for v2 or how things are done currently, so the info field will stay as-is in the near term.

Seems like an additional description field is pretty unnecessary though given all the points brought up in this thread. Just wanted to see if it would be something that could be useful, but seems like we already do/will have all the fields we need to provide context info upstream.

Thanks for the discussion and entertaining my thoughts!